### ISSUE TITLE
Operations

### ISSUE THREAD

**Issue by marccasco** [bmino/binance-triangle-arbitrage#14]
Good Morning.

This robot does not do the operations, but it could be implemented perfectly.

In which section does it detect cryptocurrencies and indicates that there is a profit?

And the APIs where they connect?

I could implement the operations if it tells me where.

Thank you.

**Comment by bmino** [comment]
Hey marccasco, thanks for your interest in this project!

You are correct that trade execution operations are not currently supported.
`Main.js` repeatedly passes arbitrage pairs to the web worker `CalculationNode.js` which performs the arbitrage calculations.

A list of the best arbitrage opportunities is maintained inside `MarketCache.js` in the private variable called `arbs` so accessing it would look like `let myList = MarketCache.arbs;`

Finally, all calls out to the Binance API are done in `BinanceApi.js` which leverages the 3rd party library `node-binance-api`

**Comment by marccasco** [comment]
Thanks for the info.
I will try to implement the operations, there will also have to add transaction fees, and deduct them from profit%.

**Comment by bmino** [comment]
I've implemented the buy/sell operation in the past. One problem I ran into was the time it takes to execute 3 trades.

Each trade took about 0.75 seconds to complete (and confirm) via api. Executing a triangular arbitrage position requires 3 trades which amounts to 2+ seconds which is a LOT of time considering how fast margins can move.

**Comment by marccasco** [comment]
The truth is that I do not know how you did the operations.

But a market-type operation takes considerably less than those 0.75 seconds.

In addition, all 3 must be done at the same time, not one after the other.

I think it's something easy to do, what worries me most would be the commissions, if we had an account with the reduced commissions to the maximum perhaps if a good benefit would be obtained.

**Comment by marccasco** [comment]
It would be to arrange X amount of all the coins. About $ 20 for each currency, and when an operation is detected that can profit, the 3 operations at once, then that is almost instantaneous at market price.

To be profitable counting the commissions, a 0.22%(0.075% vip 1) minimum difference.
That case if you are in VIP level 1, if you are at the maximum with a 0.06% difference, it would be enough to be profitable.

The problem that all currencies have to have capital to make it instantaneous.

But we could focus on 20 pairs of coins for example.

**Comment by username77** [comment]
anyone is able to post the code to place orders?
thanks

**Comment by bmino** [comment]
Once again implementing order execution is not difficult at all. The problem comes on how to do it quickly enough.

Upon some further thought, I can't come up with something better than what you seem to agree with @marccasco. We could start with a pool of X symbols that are supported (which must have initial balances) and a master symbol which the arbitrage will convert from and to.

I should have some time over the weekend to code this up

**Comment by igrekm** [comment]
Holding Alt coins in order to execute the 3 orders in parallel opens a new issue regarding the overall profitability: you would need to execute sufficient profitable arbitrages to overcome the price variation of these assets. In a bear market, that would be a killer in the long term. To mitigate this issue, the pairs could be selected based on some specific TA with short-term strong bullish performance. Not only that would reduce the risk to see the alt losing more value than the arbitrage profits, but that could be a better setup were arbitrages are more frequent (eg. strong pump on TA breakout, shaking up the orderbooks could lead to more price unbalance between the alt-btc and alt-eth pairing = more arbitrages). Adding TA, and selection of the Alts on a specific period of time complexify the full automation.

Another issue will be the trade restrictions by the exchange. Like on Binance, each coin have a restricted stepSize, tickSize and minimal notional. So you may end up having a possible arbitrage but one pair of the BTC-ALT-ETH-BTC will not have sufficient quantity to satisfy the minimal notional hence complete the 3 arbitrage steps.
This may lead as well to some Alt dust along the way and even decreasing the "master" base currency due to the lot size. Let's say you buy 10.19 ALT for 0.1 BTC, next step is ALT-ETH which has a higher lot size and the exchange will only let you trade with one decimal so only able to spend 10.1 ALT to buy ETH, leaving 0.09 ALT. If the arbitrage quantity is a small percentage, it might not be profitable anymore (increasing BTC in the previous example) if we can only rebuy ETH with less ALT than we bought from BTC. The dust could be converted later on into BNB... Once again a new parameter to take in account in the overall profitability.

**Comment by bmino** [comment]
You raise some excellent points @igrekm!

The current calculations take into consideration minimizing dust and minimum quantities. Created dust is not currently calculated into the profit calculation so there is a little extra being generated than the output suggests. For instance if the first trade obtains 10.0000005 LTC and then will exchange that for ETH, the calculation to buy ETH knows that it only has 10 LTC to purchase with.

The order book is being consulted at each tick assuming that all orders are placed at market and gobble up depth in the order book accordingly. This becomes more risky as the investment size increases because the order book will be more deeply used.

My thinking is that all orders (market orders) are placed asynchronously and profit is calculated via the difference of an account's portfolio value before any trades and after all three complete.

The long-term increase or decrease in holding value of your assets would have to be overlooked at this stage of implementation.

**Comment by marccasco** [comment]
Good Morning.
I propose to make a telegram group or similar, to discuss this topic.
I am making progress now, at the moment in the testing phase. But still I would like to form a group.

What do you think?
My telegram is @MarcCasco
If someone signs up, I create the group.

A greeting.

**Comment by bmino** [comment]
I have pushed some updates to the `order-execution` branch which implement live trading as we have discussed here. This is still in a beta stage and I am testing it with my own funds currently but I welcome some early adopters. I'd be happy to help troubleshoot and welcome PRs as we go.

I have my Discord username in my profile which I can be reached at. I can put up a room if need be.

**Comment by username77** [comment]
> I have pushed some updates to the `order-execution` branch which implement live trading as we have discussed here. This is still in a beta stage and I am testing it with my own funds currently but I welcome some early adopters. I'd be happy to help troubleshoot and welcome PRs as we go.
> 
> I have my Discord username in my profile which I can be reached at. I can put up a room if need be.

this is an good article explaining all the problems that arise using tringular arbitrage https://forum.gekko.wizb.it/thread-56746-post-57937.html

To use a telegram group is not better?

**Comment by argia-andreas** [comment]
@bmino 
Hi! So fun to play around with this! Thanks for sharing this!  :-)

Perhaps this have been mentioned already. 
I've tested the order-execution branch with parallel execution. 

I've just started looking into the code and API so perhaps I'm misunderstanding something.

One issue I ran in to was that it tried to execute the same order 3 times in a second and the orders didn't arrive or get executed in order at binance. So the balance ran out one account and the order failed as a result. 
Perhaps a way to lock and free balance when orders are reported back as executed could solve this?

**Comment by bmino** [comment]
Parallel execution requires you have sufficient balances (greater than or equal to the value specified in your config `INVESTMENT.MAX` in terms of `INVESTMENT.BASE`). This will satisfy all trades assuming that multiple arbitrage opportunities containing the same symbol are not found and executed at the same time.

I'm going to close this issue as it has gotten quite far off from the original topic.
Please feel free to open a new issue(s) with specifics!
